gradle-guide-error-prone/src/main/java/com/palantir/gradle/guide/errorprone/CallGraphBugChecker.java

```diff
@@ -0,0 +1,136 @@
+                    callGraph.addNode(calleeSymbol);
+
+                    if (caller.isPresent()) {
+                        MethodSymbol callerSymbol = caller.get();
+                        Set<MethodInvocationTree> existingInvocations =
+                                callGraph.edgeValue(callerSymbol, calleeSymbol).orElseGet(HashSet::new);
+                        existingInvocations.add(node);
+                        callGraph.putEdgeValue(callerSymbol, calleeSymbol, existingInvocations);
+                    }
+
+                    return super.visitMethodInvocation(node, caller);
+                }
+            }.scan(state.getPath().getCompilationUnit(), Optional.empty());
+
+            new SuppressibleTreePathScanner<Void, Optional<MethodInvocationTree>>(state) {
+                @Override
+                public Void visitMethodInvocation(
+                        MethodInvocationTree call, Optional<MethodInvocationTree> chainedCall) {
+                    chainedCall.ifPresent(chained -> callToChainedCall.put(call, chained));
+                    return super.visitMethodInvocation(call, Optional.of(call));
+                }
+            }.scan(state.getPath().getCompilationUnit(), Optional.empty());
+        }
+
+        /**
+         * Routine which processes an edge in the call graph.
+         * {@code callToChainedCall} maps a method call to the call immediately chained to it.
+         * We need the chained call to provide auto-fixes, e.g. getProject().getLogger() ==> getLogger()
+         */
+        public interface EdgeConsumer {
+            void accept(
+                    MethodSymbol from,
+                    MethodSymbol to,
+                    Set<MethodInvocationTree> edge,
+                    Map<MethodInvocationTree, MethodInvocationTree> callToChainedCall);
+        }
+
+        /**
+         * From `start`, visit all the methods called in `start` recursively.
+         * Uses depth-first traversal and automatically handles cycle detection.
+         *
+         * @param consumer consumes all invocations to this method
+         */
+        public void dfs(MethodSymbol start, EdgeConsumer consumer) {
+            Traverser<MethodSymbol> traverser = Traverser.forGraph(callGraph);
+
+            for (MethodSymbol current : traverser.depthFirstPreOrder(start)) {
+                for (MethodSymbol neighbor : callGraph.successors(current)) {
+                    Set<MethodInvocationTree> edgeInvocations =
+                            callGraph.edgeValue(current, neighbor).orElseGet(Set::of);
+                    consumer.accept(current, neighbor, edgeInvocations, callToChainedCall);
+                }
+            }
+        }
+    }
+
+    private static final Cache<CompilationUnitTree, MethodCallGraph> callGraphCache =
+            Caffeine.newBuilder().maximumSize(1000).weakKeys().build();
+
+    @SuppressWarnings("VisibilityModifier")
+    protected final Supplier<MethodCallGraph> callGraph =
```

Comment:
I thought I was going crazy, didn't realise the `Supplier` was an errorprone type :joy: